File: pycoder/server/routers/db_api.py
"""数据库可视化 API — F6（连接管理 / 元数据 / 查询 / ER 图）

端点:
    POST   /api/db/connect                     — 新建连接
    GET    /api/db/connections                 — 连接列表（脱敏）
    DELETE /api/db/connections/{name}          — 删除连接
    GET    /api/db/{conn}/databases            — 数据库/schema 列表
    GET    /api/db/{conn}/tables               — 表/视图列表
    GET    /api/db/{conn}/tables/{table}/schema — 表结构（列/主键/索引/外键）
    GET    /api/db/{conn}/er                   — ER 图数据（基于外键）
    POST   /api/db/{conn}/query                — 执行查询（参数化、默认只读）

安全约束:
- 所有 SQL 经 SQLAlchemy text() 绑定参数执行，禁止字符串拼接
- 默认只读：非 SELECT 语句被拒绝，除非请求显式 allow_write=true
- 连接 URL 在响应与日志中一律脱敏（密码隐藏）
"""

from __future__ import annotations

import asyncio
from typing import Any

from fastapi import APIRouter
from pydantic import BaseModel, Field

from pycoder.server.services.db_manager import (
    DatabaseError,
    ReadOnlyViolation,
    db_manager,
)

router = APIRouter(prefix="/api/db", tags=["database"])


class ConnectRequest(BaseModel):
    """新建连接请求：完整 url，或 driver + params 白名单参数"""

    name: str = Field(..., min_length=1, max_length=64)
    url: str | None = None
    driver: str | None = None
    params: dict[str, Any] | None = None


class QueryRequest(BaseModel):
    """查询请求：参数必须使用 :name 绑定变量"""

    sql: str = Field(..., min_length=1)
    params: dict[str, Any] | None = None
    limit: int | None = Field(default=None, ge=1)
    allow_write: bool = False


class ApiResponse(BaseModel):
    """统一响应格式"""

    success: bool
    data: Any = None
    error: str | None = None
    readonly: bool | None = None


def _error(exc: Exception) -> ApiResponse:
    """将服务层异常包装为统一错误响应（不泄露内部堆栈）"""
    if isinstance(exc, ReadOnlyViolation):
        return ApiResponse(success=False, error=str(exc), readonly=True)
    return ApiResponse(success=False, error=str(exc))


@router.post("/connect", response_model=ApiResponse)
async def connect(body: ConnectRequest) -> ApiResponse:
    """新建数据库连接（sqlite 直接可用；pg/mysql 缺驱动时返回明确错误）"""
    try:
        info = await asyncio.to_thread(
            db_manager.connect, body.name, body.url, body.driver, body.params
        )
        return ApiResponse(success=True, data=info.to_dict())
    except DatabaseError as exc:
        return _error(exc)


@router.get("/connections", response_model=ApiResponse)
async def list_connections() -> ApiResponse:
    """连接列表（仅脱敏信息，不含明文密码）"""
    return ApiResponse(success=True, data=db_manager.list_connections())


@router.delete("/connections/{name}", response_model=ApiResponse)
async def disconnect(name: str) -> ApiResponse:
    """断开并移除指定连接"""
    removed = await asyncio.to_thread(db_manager.disconnect, name)
    if not removed:
        return ApiResponse(success=False, error=f"连接不存在: {name!r}")
    return ApiResponse(success=True, data={"name": name})


@router.get("/{conn}/databases", response_model=ApiResponse)
async def list_databases(conn: str) -> ApiResponse:
    """列出当前连接可见的数据库/schema"""
    try:
        data = await asyncio.to_thread(db_manager.list_databases, conn)
        return ApiResponse(success=True, data=data)
    except DatabaseError as exc:
        return _error(exc)


@router.get("/{conn}/tables", response_model=ApiResponse)
async def list_tables(conn: str) -> ApiResponse:
    """列出表与视图"""
    try:
        data = await asyncio.to_thread(db_manager.list_tables, conn)
        return ApiResponse(success=True, data=data)
    except DatabaseError as exc:
        return _error(exc)


@router.get("/{conn}/tables/{table}/schema", response_model=ApiResponse)
async def describe_table(conn: str, table: str) -> ApiResponse:
    """表结构详情：列/类型/主键/索引/外键"""
    try:
        data = await asyncio.to_thread(db_manager.describe_table, conn, table)
        return ApiResponse(success=True, data=data)
    except DatabaseError as exc:
        return _error(exc)


@router.get("/{conn}/er", response_model=ApiResponse)
async def table_relationships(conn: str) -> ApiResponse:
    """ER 图数据：nodes（表+列）与 edges（外键关系）"""
    try:
        data = await asyncio.to_thread(db_manager.table_relationships, conn)
        return ApiResponse(success=True, data=data)
    except DatabaseError as exc:
        return _error(exc)


@router.post("/{conn}/query", response_model=ApiResponse)
async def execute_query(conn: str, body: QueryRequest) -> ApiResponse:
    """执行 SQL 查询（强制参数化、默认只读、行数上限、执行计时）"""
    try:
        result = await asyncio.to_thread(
            db_manager.execute_query,
            conn,
            body.sql,
            body.params,
            body.limit,
            body.allow_write,
        )
        return ApiResponse(success=True, data=result.to_dict())
    except (DatabaseError, ReadOnlyViolation) as exc:
        return _error(exc)

File: pycoder/electron/src/renderer/services/dbApi.ts
/**
 * 数据库可视化 API 客户端 — F6
 *
 * 通过 fetch 调用后端 `/api/db/*` 系列端点。
 * 所有方法返回统一响应 { success, data, error }，请求失败时返回 null。
 *
 * 后端路由源: pycoder/server/routers/db_api.py
 */

import { getApiBase, getApiKey } from './config';

// ═══════════════════════════════════════════════════════════
// 类型定义
// ═══════════════════════════════════════════════════════════

/** 连接信息（url 已脱敏，不含明文密码） */
export interface DbConnection {
    name: string;
    driver: string;
    url: string;
    created_at: number;
}

/** 表/视图列表项 */
export interface DbTable {
    name: string;
    type: 'table' | 'view';
}

/** 列定义 */
export interface DbColumn {
    name: string;
    type: string;
    nullable: boolean;
    default: string | null;
    primary_key: boolean;
}

/** 索引定义 */
export interface DbIndex {
    name: string | null;
    columns: string[];
    unique: boolean;
}

/** 外键定义 */
export interface DbForeignKey {
    columns: string[];
    referred_table: string;
    referred_columns: string[];
}

/** 表结构详情 */
export interface DbTableSchema {
    table: string;
    columns: DbColumn[];
    primary_key: string[];
    indexes: DbIndex[];
    foreign_keys: DbForeignKey[];
}

/** 查询执行结果 */
export interface DbQueryResult {
    columns: string[];
    rows: unknown[][];
    row_count: number;
    elapsed_ms: number;
    truncated: boolean;
}

/** ER 图节点（表 + 列） */
export interface ErNode {
    name: string;
    columns: Array<{ name: string; type: string; primary_key: boolean }>;
}

/** ER 图边（外键关系） */
export interface ErEdge {
    from_table: string;
    from_columns: string[];
    to_table: string;
    to_columns: string[];
}

/** ER 图数据 */
export interface ErData {
    nodes: ErNode[];
    edges: ErEdge[];
}

/** 统一响应格式 */
export interface DbApiResponse<T = unknown> {
    success: boolean;
    data?: T;
    error?: string;
    /** true 表示被只读模式拒绝 */
    readonly?: boolean;
}

// ═══════════════════════════════════════════════════════════
// 请求封装
// ═══════════════════════════════════════════════════════════

async function request<T = unknown>(
    path: string,
    options: RequestInit = {},
): Promise<DbApiResponse<T> | null> {
    try {
        const controller = new AbortController();
        const timeout = setTimeout(() => controller.abort(), 60000);
        const base = await getApiBase();
        const apiKey = await getApiKey();
        const headers: Record<string, string> = {
            ...((options.headers as Record<string, string>) || {}),
        };
        if (options.body && !headers['Content-Type']) {
            headers['Content-Type'] = 'application/json';
        }
        if (apiKey) {
            headers['X-API-Key'] = apiKey;
        }
        const res = await fetch(`${base}${path}`, {
            ...options,
            headers,
            signal: controller.signal,
        });
        clearTimeout(timeout);
        return (await res.json()) as DbApiResponse<T>;
    } catch (err) {
        if ((err as Error).name !== 'AbortError') {
            console.error(`[dbApi] ${path} failed:`, err);
        }
        return null;
    }
}

// ═══════════════════════════════════════════════════════════
// API 方法
// ═══════════════════════════════════════════════════════════

export const dbApi = {
    /** 新建连接：SQLite 传 { name, driver: 'sqlite', params: { path } }，或完整 url */
    connect(payload: {
        name: string;
        url?: string;
        driver?: string;
        params?: Record<string, unknown>;
    }) {
        return request<DbConnection>('/api/db/connect', {
            method: 'POST',
            body: JSON.stringify(payload),
        });
    },

    /** 连接列表（脱敏） */
    listConnections() {
        return request<DbConnection[]>('/api/db/connections');
    },

    /** 删除连接 */
    disconnect(name: string) {
        return request<{ name: string }>(
            `/api/db/connections/${encodeURIComponent(name)}`,
            { method: 'DELETE' },
        );
    },

    /** 数据库/schema 列表 */
    listDatabases(conn: string) {
        return request<string[]>(`/api/db/${encodeURIComponent(conn)}/databases`);
    },

    /** 表/视图列表 */
    listTables(conn: string) {
        return request<DbTable[]>(`/api/db/${encodeURIComponent(conn)}/tables`);
    },

    /** 表结构（列/主键/索引/外键） */
    describeTable(conn: string, table: string) {
        return request<DbTableSchema>(
            `/api/db/${encodeURIComponent(conn)}/tables/${encodeURIComponent(table)}/schema`,
        );
    },

    /** ER 图数据（基于外键） */
    getEr(conn: string) {
        return request<ErData>(`/api/db/${encodeURIComponent(conn)}/er`);
    },

    /** 执行查询（参数化、默认只读） */
    query(
        conn: string,
        payload: {
            sql: string;
            params?: Record<string, unknown>;
            limit?: number;
            allow_write?: boolean;
        },
    ) {
        return request<DbQueryResult>(
            `/api/db/${encodeURIComponent(conn)}/query`,
            { method: 'POST', body: JSON.stringify(payload) },
        );
    },
};
